[analyzer][docs] Document the `optin.performance.Padding` checker

komalverma04 opened the pull request:
# Added documentation for optin.performance.Padding.
- Performance package has `PaddingChecker` checker.
- It checks for excessively padded structs.
   - It has one option that is `AllowedPad`, an integer option.
   - It's default value is 24 bytes.
      - Reports are generated when padding exceeds `AllowedPad`.
 ## Purpose
The purpose of this pull request is to improve the clarity and completeness of the documentation for PaddingChecker in the optin.performance.Padding checker.

Closes #73675

komalverma04 committed:
Update checkers.rst

Modification of documentation to demonstrate utilization of AllowedPad within PaddingChecker, along with its use and effects on code analysis.

komalverma04 committed:
Update Checkers.td

Changed NotDocumented to HasDocumentation for Padding Checker under performance checker.

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,28 @@ Check for performance anti-patterns when using Grand Central Dispatch.
 
-optin.performance.Padding
-"""""""""""""""""""""""""
+optin.performance.Padding (PaddingChecker)
```
It should be the languages within the parens, that the checker supports.
I think this checker applies for C, C++ and probably even for ObjC as well.

komalverma04 replied:
Thank you so much for your valuable feedbacks i will update it as you said.

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,28 @@ Check for performance anti-patterns when using Grand Central Dispatch.
-optin.performance.Padding
-"""""""""""""""""""""""""
+optin.performance.Padding (PaddingChecker)
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```
This line should be aligned with the line above.

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,28 @@ Check for performance anti-patterns when using Grand Central Dispatch.
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
 Check for excessively padded structs.
 
+.. code-block:: objc
```
This code is rather c.

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,28 @@ Check for performance anti-patterns when using Grand Central Dispatch.
+ struct TestStruct {
+      int data1;   // 4 bytes
+      char data2;  // 1 byte
+      char padding[27];  // 27 bytes of padding
```
This is a field, thus it shouldn't be counted as padding.

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,28 @@ Check for performance anti-patterns when using Grand Central Dispatch.
+      char padding[27];  // 27 bytes of padding
+    };  // Total size: 32 bytes 
+  
+  void TestFunction() {
```
All lines in the example block should be indented nicely in a consistent manner.
As of now, the indentation level varies.

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,28 @@ Check for performance anti-patterns when using Grand Central Dispatch.
+
+  // AllowedPad: Default Value: 24 bytes
+
+   Usage: `AllowedPad=<value>`
```
In other places we use double backticks. Why do you use one where?

steakhal reviewed (changes requested):
Thank you for working on documenting this checker. Docs are a really neglected within CSA.

The docs should also explain why is a problem to have padding bytes, and how to fix it.

How does the generated html look like after your change for this file?
(Have a look at my commits to this or to the releasenotes file for clang to see what commands to run to generate it)
Like [here](https://github.com/llvm/llvm-project/pull/76446#issuecomment-1870420291).

Does the checker have any limitations? Why is it even implemented in CSA?

komalverma04 committed:
Update checkers.rst

Made Indentation consistent and added description on limitation of checker.

komalverma04 committed:
Update checkers.rst

Fixed some typos

AZero13 reviewed (changes requested):
Capitalize the O in ObjC

komalverma04 committed:
Update checkers.rst

Fixed a typo in objC and changed it to ObjC

komalverma04 commented:
@AtariDreams Thank you for pointing out. I apologies for that mistake. Please review it, if there are more changes.

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,38 @@ Check for performance anti-patterns when using Grand Central Dispatch.
+
+Reports are only generated if the excessive padding exceeds ``AllowedPad`` in bytes. AllowedPad is the threshold value of padding.
+
+AllowedPad Option:
+- Default Value: 24 bytes
+- Usage: ``AllowedPad=<value>``
```
I think you could just drop this, and just mention the default value in the previous paragraph.
Maybe add a note like:
> To enable set this option from the command-line, pass the `-analyzer-config optin.performance.Padding:AllowedPad=24` option.

komalverma04 replied:
Okay, i will do the changes as you said.

komalverma04 replied:
@steakhal @AtariDreams Please review it once. Thank you.

steakhal reviewed (changes requested):
The grammar seems good.
I have further concerns inline.

komalverma04 commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,38 @@ Check for performance anti-patterns when using Grand Central Dispatch.
+Padding can be disabled by using the pragma directive.
+Padding can also be decreased by putting data members of the structure in descending order of their size.
+
+Reports are only generated if the excessive padding exceeds ``AllowedPad`` in bytes. AllowedPad is the threshold value of padding.
```
I will write everything in described form.

komalverma04 replied:
@steakhal, if fields are in optimal order that is widest datatype first,still If the padding between two fields exceeds this value, the analyzer will issue a warning indicating that there's excessive padding within the structure.
Okay, so i will write it in the documentation in described form.
Please correct me if i am wrong

komalverma04 committed:
Add description for ``AllowedPad``, fix space inconsistency and modify the example code

Fixing of trailing spaces. Described the option to make more clarity and modify the example code to correctly demonstrate the issue and added improved code also for more clarity.

steakhal committed:
Rework docs

steakhal commented:
I figured, it might be easier to directly propose and apply some of my thoughts.
Let me know if you like it.

My goal was:
 - be specific what is the impact of not having the optimal layout: memory, cachemisses
 - have a single block of examples, demonstrating the bad, the good and an alternative good (bitpacking) case, while have no header dependencies
 - rework the example to work with the default configuration of the checker (this makes it more similar to the average user's diagnostics)
 - move all the advanced explanations to the end, as most users won't consider setting a custom threshold
 - make the definition of "when it warns" more accurate. I've checked [the code](https://github.com/llvm/llvm-project/blob/main/clang/lib/StaticAnalyzer/Checkers/PaddingChecker.cpp#L103-L109) and it seems that the difference counts compared to the optimal layout.
 - discourage the use of `pragma pack`
 - add guide for fixing the issue, along with a practical hint.
 
 Given that it looks now as I would like it, I'll invite @NagyDonat for a round of review.
 Let me know if you like it too, or have suggestions for making it better.
 
 Lastly, now the generated html would look like this:
 
![image](https://github.com/llvm/llvm-project/assets/6280485/9ea66343-5713-4b7d-b365-e81ae1ebb39e)

NagyDonat commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,88 @@ Check for performance anti-patterns when using Grand Central Dispatch.
+ordering the fields by ``sizeof`` would usually also lead to the same optimal
+layout.
+
+In rare cases, one can use the ``#pragma pack(1)`` directive to enforce a packed
+layout too, but it is discouraged and the reordering of fields should be
+preferred.
```
```suggestion
In rare cases, one can use the ``#pragma pack(1)`` directive to enforce a packed
layout too, but it can significantly increase the access times, so reordering the
fields is usually a better solution.

```

NagyDonat commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -804,10 +804,88 @@ Check for performance anti-patterns when using Grand Central Dispatch.
+ };
+ static_assert(sizeof(Optimal) == 4*8+5+3);
+
+ // no-warning: Enforcing bitpacked representation.
+ // Access times will have signifficant overhead. Prefer reordering the fields instead.
```
```suggestion
 // no-warning: Bit packing representation is also accepted by this checker, but
 // it can significantly increase access times, so prefer reordering the fields.
```

NagyDonat reviewed (approved):
Overall LGTM, I have minor inline suggestions that clarify the description of bitpacking (but the current state is also acceptable).

komalverma04 commented:
Okay, i completely agree with you.
Thank you so much for helping.

steakhal committed:
Update clang/docs/analyzer/checkers.rst

Co-authored-by: NagyDonat <donat.nagy@ericsson.com>

steakhal committed:
Update clang/docs/analyzer/checkers.rst

Co-authored-by: NagyDonat <donat.nagy@ericsson.com>

komalverma04 reviewed (commented):
Great learning!
